examples/filter_data.py

```diff
@@ -160,20 +175,60 @@ def main(
             pass
         elif args.filter == "attribution":
             if query == None:
-                query = train["gradient"].mean(0).cuda()
+                # Calculate mean gradient in batches to avoid memory issues
+                print("Calculating mean gradient...")
+                mean_gradient = None
+                batch_size = 100  # Process in smaller batches
+                
+                for i in tqdm(range(0, len(train), batch_size), desc="Computing mean gradient"):
+                    batch = train.select(range(i, min(i + batch_size, len(train))))
+                    batch_gradients = torch.stack([g for g in batch["gradients"]])
+                    
+                    if mean_gradient is None:
+                        mean_gradient = batch_gradients.mean(0)
+                    else:
+                        mean_gradient = (mean_gradient * i + batch_gradients.sum(0)) / (i + len(batch))
+                    
+                    # Clear memory
+                    del batch_gradients
+                
+                query = mean_gradient.cuda()
             
-            def importance_score_generator():
-                for row in train:
-                    row = dict(row)
-                    yield {
-                        **row,
-                        "importance_score": (row["gradient"].cuda() @ query).item(),
-                    }
-
-            train = assert_type(
-                Dataset, Dataset.from_generator(importance_score_generator)
-            )
-            train = select_topk(train, args.num_examples, "importance_score", lowest=args.lowest)
+            print("Computing importance scores...")
+            # Calculate importance scores in batches
+            importance_scores = []
+            batch_size = 512
+            
+            for i in tqdm(range(0, len(train), batch_size), desc="Computing importance scores"):
+                batch = train.select(range(i, min(i + batch_size, len(train))))
```

Comment:
I _think_ you can do `train[i : min(i + batch_size, len(train))]` then remove the `torch.stack`